{{cookiecutter.project_name}}/{{cookiecutter.package_name}}/__init__.py

```diff
@@ -3,93 +3,117 @@
 import json
 import singer
 from singer import utils, metadata
+from singer.catalog import Catalog, CatalogEntry
+from singer.schema import Schema
 
+
+DEFAULT_BATCH_SIZE = 1000
 REQUIRED_CONFIG_KEYS = ["start_date", "username", "password"]
 LOGGER = singer.get_logger()
 
+
 def get_abs_path(path):
     return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)
 
-# Load schemas from schemas folder
+
 def load_schemas():
+    """ Load schemas from schemas folder """
     schemas = {}
-
     for filename in os.listdir(get_abs_path('schemas')):
         path = get_abs_path('schemas') + '/' + filename
         file_raw = filename.replace('.json', '')
         with open(path) as file:
-            schemas[file_raw] = json.load(file)
-
+            schemas[file_raw] = Schema.from_dict(json.load(file))
     return schemas
 
+
 def discover():
     raw_schemas = load_schemas()
     streams = []
-
     for schema_name, schema in raw_schemas.items():
-
         # TODO: populate any metadata and stream's key properties here..
         stream_metadata = []
-        stream_key_properties = []
-
-        # create and add catalog entry
-        catalog_entry = {
-            'stream': schema_name,
-            'tap_stream_id': schema_name,
-            'schema': schema,
-            'metadata' : [],
-            'key_properties': []
-        }
-        streams.append(catalog_entry)
-
-    return {'streams': streams}
-
-def get_selected_streams(catalog):
-    '''
-    Gets selected streams.  Checks schema's 'selected' first (legacy)
-    and then checks metadata (current), looking for an empty breadcrumb
-    and mdata with a 'selected' entry
-    '''
-    selected_streams = []
-    for stream in catalog.streams:
-        stream_metadata = metadata.to_map(stream.metadata)
-        # stream metadata will have an empty breadcrumb
-        if metadata.get(stream_metadata, (), "selected"):
-            selected_streams.append(stream.tap_stream_id)
+        key_properties = []
+        streams.append(
+            CatalogEntry(
+                tap_stream_id=schema_name,
+                stream=schema_name,
+                schema=schema,
+                key_properties=key_properties,
+                metadata=stream_metadata,
+                replication_key=None,
+                is_view=None,
+                database=None,
+                table=None,
+                row_count=None,
+                stream_alias=None,
+                replication_method=None,
+            )
+        )
+    return Catalog(streams)
+
+
+def read_all_data():
+    """ Read all rows from upstream source and return data in proper json format """
+    # TODO: Replace with real logic to retrieve data
+    for dummy_value in range(10000):
+        yield {"id": dummy_value, "name": str(dummy_value)}
+
+
+def get_row_batches(batch_size=1):
```

Comment:
I'm a bit hesitant to add a concept like this in here, since it might be too opinionated for all source types. I've seen a lot of different implementations that use things like `paginate`, a `client` object, etc. so it might be good to leave it out.

Reply:
Hmm, though I do like the `singer.write_records` call being in the sync function below. Maybe just a doc string or comment that mentions that this could be replaced with whichever implementation makes the most sense?

Reply:
@dmosorast - I agree with that feedback about wanting to reduce influence of opinionated implementations. That said, and in line with your second comment, my primary objective was to provide sample code for `write_records()`, `write_schema()`, and `write_stage()` since those were all in the readme doc but not referenced here in the doc.

What if I take another stab at this using a simple dummy dictionary iterator inside `sync()` - without the `get_data()` function and the batching function?

something like:

```python
    def sync(...):
        ...
        #TODO: delete and replace this inline function with your own data retrieval process:
        tap_data = lambda: [{"id": x, name: "row${x}"} for x in range(1000)]
        ...
        for row in tap_data():
            # TODO: place any necessary transformations here

            # write one or more rows to the stream:
            singer.write_records(stream_id, [row]) 
        ...
```

If possible, I'd like to retain at least one call to `singer.write_records()` and I think it makes sense for it to be within some type of loop or for construct. What do you think?

Reply:
@aaronsteers That sounds like a good approach!

Reply:
@dmosorast - great, thanks. I've committed an update that takes this approach; I've removed the extra get data functions and replaced with the simple inline lambda (with comment explaining).